# Comparative Gene Coexpression Analysis between Sugarcane and Sorghum

This repository contains scripts and data for the comparative gene coexpression analysis between sugarcane and sorghum. The project aims to identify and compare gene coexpression networks in these two important crop species.

## Table of Contents
- [Introduction](#introduction)
- [Data](#data)
- [Methods](#methods)
- [Results](#results)
- [Installation](#installation)
- [Usage](#usage)
- [Contributing](#contributing)
- [License](#license)
- [Acknowledgements](#acknowledgements)

## Introduction
Gene coexpression analysis is a powerful tool for understanding the functional relationships between genes. This project focuses on comparing gene coexpression networks in sugarcane and sorghum to identify conserved and species-specific coexpression patterns in the sugar-fiber accumulation proccess.

## Data
The data used in this analysis include:
- Gene expression datasets retrieved from SRA (https://www.ncbi.nlm.nih.gov/sra) for Sugarcane and Sorghum. for Sorghum  it includes 115 accessions and six genotypes, three grain and three sweet. For Sugarcane it includes 109 accessions including seven gentypes, three rich in sugar and four rich in fiber, metadata in data/sugarcane_metadata.csv and data/sorghum_metadata.csv

- Sorghum Pantranscriptome (Transcript of longest cds per orthogroup):
- Sugarcane Pantranscriptome (Transcript of longest cds per orthogroup):


| Genotype | Number of Accessions | Group | Species |
|----------|----------------------|-------|---------|
| BTx623   | 20                   |sweet  |Sorghum  |
| RTx430   | 20                   |sweet  |Sorghum  |
| BTx642   | 20                   |sweet  |Sorghum  |
| della    | 23                   |grain  |Sorghum  |
| keller   | 20                   |grain  |Sorghum  |
| rio      | 12                   |grain  |Sorghum  |
| SRA5     | 15                   |sugar  |Sugarcane|
| SRA3     | 15                   |sugar  |Sugarcane|
| Q238     | 15                   |sugar  |Sugarcane|
| KQB09-20432| 19                 |fiber  |Sugarcane|
| Q157     | 15                   |fiber  |Sugarcane|
| Q151     | 15                   |fiber  |Sugarcane|
| Q135     | 15                   |fiber  |Sugarcane|

complete list of accession in accessions directory

## Methods
The analysis was performed using the following steps:
1. Preprocessing of gene expression data.
    - Download sequences using sra-toolkit (sra-tools 2.10.0)
    - bbduk: remove poor quality sequences short sequences and remove rRNA (bbmap 38.84)
    - Read mapping salmon (v1.10.2) in Sorghum and Sugarcane pantranscriptomes references (transcript of longest cds per orthogroup) ( salmon 0.12.0)

There is a snakefile for this steps in snakefile_preprocessing.py, download_sra.sh launches snakefile in SLURM type system.
plot_quant.r is useful to take the results of get_results.sh and plot the quantfification data
get_bbduk_results.sh is useful to join bbduk results
plot_bbduk can plot the results from get_bbduk_results.sh



2. Data Quantification
    - salmon (v1.10.2)ata and plot exploratory PCA of vst transformed data (pca_for_rnaseq.r)
    - get_results.sh collect salmon results
    - plot_quant.r plot salmon results

3. Plot PCA
    - pca_for_rnaseq.r plot PCA using DESeq2

4. Correct for batch effects
    - RUV.r implements RUVr (https://bioconductor.org/packages/release/bioc/vignettes/RUVSeq/inst/doc/RUVSeq.html) method and plot PCA  	 

5. Make network
    - pearson_cor.r ultra fast pearson correlation method in R using HiClimR (https://cran.r-project.org/web/packages/HiClimR/index.html)
    - histogram_cor.py make plot of the distribution of the correlations
    - very_fancy_python_code_for_melt.py select the correlations greater than 0.6 and makes an edge list from the pearson correlations matrix (multicore, very fast and memory efficient)

6. Find modules in network
    - run_mcl.sh run mcl (i=2.7) edge list generated by very_fancy_python_code_for_melt.py
    - format_modules.sh format mcl output in a convinient way (gene,module)

7. Correlation analysis of modules
    - corr_modules.r calculate the correlation of the eigengene of each module with the sugar-fiber trait

8. Annotate data
    - run_Panzzer2.sh assing GO to proteins
    - run_tf.sh run hmmer to find TF proteind domais
    - assign_family_membership.pl make table with tf annotations from run_tf.sh

9. Get orthologues
    - run_orthofinder.sh find othologues with orthofinder (2.5.4) on sugcarcane, sorghum, maize and rice proteins.

10. Make heatmaps of expression data
    - make_heatmaps_for_all_modules.r make heatmaps for all modules and for module eigengenes correlated to sugar-fiber trait
    - make heatmap of orthologues genes in modules correlated to sugar-fiber trait

11. Enrichment GO analysis
    - topGO.r makes overrepresentation test on each module
    - compare_modules.r make overrepresentation analyisis of groups genes of modules correlated to sugar-fiber trait

12. Compare the correlated modules in sugarcane and sorghum
    - compare_modules.r Detect orthologues genes in correlated modules in sugarcane and sorghum
    - compare_modules.r Detect transcription factors in orthologues genes correlated to sugar-fiber trait in sorghum and sugarcane

## Results

We download a total of 14122223702 raw sequences, 6651867490 from sorghum and 7470356212 from sugarcane, after removing poor quality, rRNA, very short sequences and mapping with salmon against pan-transcriptome references we got 10698352940 sequences (75.6% of total download sequences, 84.5% sorghum and 68.0% for sugarcane), see table and plots in bbduk results. 

Sorghum quantification
![alt text](https://github.com/jomare1188/glowing-palm-tree/blob/master/results/quant/quant_sorghum.png?raw=true)

Sugarcane quantification
![alt text](https://github.com/jomare1188/glowing-palm-tree/blob/master/results/quant/quant_scane.png?raw=true)

Raw PCA for sorghum
![alt text](https://github.com/jomare1188/glowing-palm-tree/blob/master/results/PCA/pca_sorghum_group.png?raw=true)

Raw PCA For sugarcane
![alt text](https://github.com/jomare1188/glowing-palm-tree/blob/master/results/PCA/pca_sugarcane_group.png?raw=true)

Raw facet PCA for sugarcane
![alt text](https://github.com/jomare1188/glowing-palm-tree/blob/master/results/PCA/pca_sugarcane_facet.png?raw=true)

Normalized PCA for sorghum - k1
![alt text](https://github.com/jomare1188/glowing-palm-tree/blob/master/results/RUVr/sorghum_k1_RUVr_groups.png?raw=true)

Normalized PCA for sugarcane - k2
![alt text](https://github.com/jomare1188/glowing-palm-tree/blob/master/results/RUVr/sugarcane_k2_RUVr_groups.png?raw=true)

Count matrixes: Matrices directory 
## Installation
You can find yml configuration files for conda envs in conda_envs directory
